lib/server.js

```diff
@@ -55,22 +55,36 @@ function createSocketIoServer (webServer, executor, config) {
 }
 
 class Server extends KarmaEventEmitter {
-  constructor (cliOptions, done) {
+  constructor (cliOptionsOrConfig, done) {
     super()
-    logger.setupFromConfig(cliOptions)
-
+    cliOptionsOrConfig = cliOptionsOrConfig || {}
+    logger.setupFromConfig({
+      colors: cliOptionsOrConfig.colors,
+      logLevel: cliOptionsOrConfig.logLevel
+    })
     this.log = logger.create('karma-server')
-
+    done = helper.isFunction(done) ? done : process.exit
     this.loadErrors = []
 
     let config
-    try {
-      config = cfg.parseConfig(cliOptions.configFile, cliOptions, { throwErrors: true })
-    } catch (parseConfigError) {
-      // TODO: change how `done` falls back to exit in next major version
-      //  SEE: https://github.com/karma-runner/karma/pull/3635#discussion_r565399378
-      (done || process.exit)(1)
-      return
+    if (cliOptionsOrConfig instanceof cfg.Config) {
+      config = cliOptionsOrConfig
+    } else {
+      try {
```

Comment:
Can you also log a deprecation warning in the `else` branches in server, runner and stopper?

Something like:

Passing CLI options to the {component} is deprecated and will be removed in the next major release. Use `parseConfig(..., {throwError: true, promiseConfig: true})` to prepare an instance of the `Config` class and pass it instead of raw CLI options.